base/intfuncs.jl

```diff
@@ -602,74 +602,111 @@ ndigits(x::Integer; base::Integer=10, pad::Integer=1) = max(pad, ndigits0z(x, ba
 ## integer to string functions ##
 
 function bin(x::Unsigned, pad::Integer, neg::Bool)
-    i = neg + max(pad,sizeof(x)<<3-leading_zeros(x))
-    a = StringVector(i)
+    m = 8sizeof(x) - leading_zeros(x)::Int
+    n = neg + max((pad % Int)::Int, m)
```

Comment:
Is the type assertion really necessary in `(pad % Int)::Int`? That seems like something the compiler should know.

Reply:
> Is the type assertion really necessary in (pad % Int)::Int

In some circumstances, yes. All applicable methods infer to `Int` but there are 14 of them:

```julia
julia> sig = (Integer, Type{Int})
(Integer, Type{Int64})

julia> cts = code_typed(%, sig);

julia> collect(zip(methods(%, sig), last.(cts)))
14-element Array{Tuple{Method,DataType},1}:
 (rem(x::BigInt, ::Type{T}) where T<:Union{Int16, Int32, Int64, Int8, UInt16, UInt32, UInt64, UInt8} in Base.GMP at gmp.jl:332, Int64)
 (rem(x::BigInt, ::Type{T}) where T<:Union{Int128, Int16, Int32, Int64, Int8, UInt128, UInt16, UInt32, UInt64, UInt8} in Base.GMP at gmp.jl:335, Int64)
 (rem(x::UInt32, ::Type{Int64}) in Base at int.jl:479, Int64)
 (rem(x::Int8, ::Type{Int64}) in Base at int.jl:477, Int64)
 (rem(x::Bool, ::Type{Int64}) in Base at int.jl:474, Int64)
 (rem(x::UInt16, ::Type{Int64}) in Base at int.jl:479, Int64)
 (rem(x::UInt128, ::Type{Int64}) in Base at int.jl:472, Int64)
 (rem(x::UInt64, ::Type{Int64}) in Base at int.jl:482, Int64)
 (rem(x::UInt8, ::Type{Int64}) in Base at int.jl:479, Int64)
 (rem(x::Int128, ::Type{Int64}) in Base at int.jl:472, Int64)
 (rem(x::Int32, ::Type{Int64}) in Base at int.jl:477, Int64)
 (rem(x::Int16, ::Type{Int64}) in Base at int.jl:477, Int64)
 (rem(x::T, ::Type{T}) where T<:Integer in Base at int.jl:540, Int64)
 (rem(x::Integer, T::Type{var"#s71"} where var"#s71"<:Integer) in Base at int.jl:543, Int64)
```
and `max_methods` is 3. So the abstractly-inferred call is judged to have a return type of `Any`, even if any concrete type is inferred correctly:
```julia
julia> f(x) = x % Int
f (generic function with 1 method)

julia> code_typed(f, (Int,))
1-element Array{Any,1}:
 CodeInfo(
1 ─     return x
) => Int64

julia> code_typed(f, (Integer,))
1-element Array{Any,1}:
 CodeInfo(
1 ─ %1 = (x % Main.Int)::Any
└──      return %1
) => Any
```

Because we are discovering more and more good reasons *not* to run inference to exhaustion (and developing more infrastructure to limit inference, like `@nospecialize`), this is becoming more of an issue. This is basically what #36463 is about.

Reply:
Gotcha. Carry on.